ios/browser/api/history/brave_history_api.mm

```diff
@@ -0,0 +1,309 @@
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  history::HistoryAddPageArgs args;
+  args.url = net::GURLWithNSURL(history.url);
+  args.time =
+      base::Time::FromDoubleT([history.dateAdded timeIntervalSince1970]);
+  args.redirects = history::RedirectList();
+  // Important! Only Typed Transtion is being synced in core side
+  args.transition = pageTransition == BraveHistoryTransitionType_TYPED
+                        ? ui::PAGE_TRANSITION_TYPED
+                        : ui::PAGE_TRANSITION_LINK;
+  args.hidden = false;
+  args.visit_source = history::VisitSource::SOURCE_BROWSED;
+  args.consider_for_ntp_most_visited = true;
+  args.title = base::SysNSStringToUTF16(history.title);
+  args.floc_allowed = false;  // Disable Floc - Not allow tracking
+
+  history_service_->AddPage(args);
+}
+
+- (void)removeHistory:(IOSHistoryNode*)history {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  // Deletes a specific URL using history service and web history service
+  history_service_->DeleteLocalAndRemoteUrl(web_history_service_,
+                                            net::GURLWithNSURL(history.url));
+}
+
+- (void)removeAllWithCompletion:(void (^)())completion {
+  DCHECK_CURRENTLY_ON(web::WebThread::UI);
+  DCHECK(history_service_->backend_loaded());
+
+  // Deletes entire History and from all synced devices
+  history_service_->DeleteLocalAndRemoteHistoryBetween(
+      web_history_service_, base::Time::Min(), base::Time::Max(),
+      base::BindOnce(completion), &tracker_);
+}
+
+- (void)searchWithQuery:(NSString*)query
+               maxCount:(NSUInteger)maxCount
+             completion:(void (^)(NSArray<IOSHistoryNode*>* historyResults))
+                            completion {
+  __weak BraveHistoryAPI* weak_history_api = self;
+
+  auto search_with_query = ^(
+      NSString* query, NSUInteger maxCount,
+      void (^completion)(NSArray<IOSHistoryNode*>* historyResults)) {
+    BraveHistoryAPI* history_api = weak_history_api;
+    if (!history_api) {
+      completion(@[]);
+      return;
+    }
+
+    DCHECK(history_api->history_service_->backend_loaded());
+
+    // Check Query is empty for Fetching all history
+    // The entered query can be nil or empty String
+    BOOL fetchAllHistory = !query || [query length] == 0;
+    std::u16string queryString =
+        fetchAllHistory ? std::u16string() : base::SysNSStringToUTF16(query);
+
+    // Creating fetch options for querying history
+    history::QueryOptions options;
+    options.duplicate_policy =
+        fetchAllHistory ? history::QueryOptions::REMOVE_DUPLICATES_PER_DAY
+                        : history::QueryOptions::REMOVE_ALL_DUPLICATES;
+    options.max_count = fetchAllHistory ? 0 : static_cast<int>(maxCount);
+    options.matching_algorithm =
+        query_parser::MatchingAlgorithm::ALWAYS_PREFIX_SEARCH;
+
+    history_api->_query_completion = completion;
```

Comment:
this isn't safe, if you make a second query before the first one returns you'll send the results to the wrong callback

Reply:
The `BraveBrowsingHistoryDriverIOS` class is removed. The `searchQuery` is re-written using HistoryService.

At this moment we dont have pagination on iOS side for history, to support this client side implementation for the UI and data model also has to change. 

We will add our own `HistoryConsumer` on core part when pagination will be implemented on mobile side.

Fix at https://github.com/brave/brave-core/pull/8869/commits/d3eca9d7a0816f889c22576e57c1b0d32282f02e